jbpm-flow/src/main/java/org/jbpm/process/core/context/variable/CaseVariableInstance.java

```diff
@@ -0,0 +1,102 @@
+import org.drools.core.common.InternalKnowledgeRuntime;
+import org.jbpm.process.instance.context.variable.VariableScopeInstance;
+import org.kie.api.runtime.KieSession;
+import org.kie.api.runtime.process.CaseData;
+import org.kie.api.runtime.rule.FactHandle;
+
+public class CaseVariableInstance<T> implements VariableInstance<T> {
+
+    private final InternalKnowledgeRuntime knowledgeRuntime;
+    private final Variable variable;
+    private final String caseFileName;
+    private final ValueReference<T> valueReference = new ValueReference<T>() {
+        @Override
+        public T get() {
+            Collection<CaseData> caseFiles = (Collection<CaseData>)
+                    knowledgeRuntime.getObjects(new ClassObjectFilter(CaseData.class));
+            if (caseFiles.size() == 1) {
+                CaseData caseFile = caseFiles.iterator().next();
+                // check if there is case file prefix and if so remove it before checking case file data
+                if (caseFile != null) {
+                    return (T) caseFile.getData(caseFileName);
+                }
+            }
+            return null;
+        }
+
+        @Override
+        public void set(T value) {
+            Collection<CaseData> caseFiles = (Collection<CaseData>)
+                    knowledgeRuntime.getObjects(new ClassObjectFilter(CaseData.class));
+            if (caseFiles.size() == 1) {
+                CaseData caseFile = caseFiles.iterator().next();
+                FactHandle factHandle = knowledgeRuntime.getFactHandle(caseFile);
+                Objects.requireNonNull(factHandle, "factHandle for " + caseFileName + " was null");
+                caseFile.add(caseFileName, value);
+                knowledgeRuntime.update(factHandle, caseFile);
+                ((KieSession) knowledgeRuntime).fireAllRules();
+            }
+        }
+    };
+
+    public CaseVariableInstance(VariableScopeInstance parentScopeInstance, Variable variable) {
+        this.knowledgeRuntime = parentScopeInstance.getProcessInstance().getKnowledgeRuntime();
+        this.variable = variable;
+        if (variable.getName().startsWith(VariableScope.CASE_FILE_PREFIX)) {
+            this.caseFileName = variable.getName().substring(VariableScope.CASE_FILE_PREFIX.length());
+        } else {
+            // should probably issue a warning
+            this.caseFileName = variable.getName();
+        }
+    }
+
+    @Override
+    public T get() {
+        return valueReference.get();
+    }
+
+    @Override
+    public void set(T value) {
+        valueReference.set(value);
+    }
+
+    @Override
+    public void setReference(ValueReference<T> value) {
+        // shall we log this? "Cannot setReference on a CaseVariableInstance". will set(value.get())
+        set(value.get());
+    }
+
+    @Override
+    public ValueReference<T> getReference() {
+        // fixme: is a CaseVariable really an InstanceVariable? Shall we use a ValueReference?
```

Comment:
I’d say it should be InstanceVariable, the main difference is the scope of it - case variable is broader than process variable

Reply:
this is rather a question on the class hierarchy/class relations here.

A CaseVariable here is a subclass of an InstanceVariable; but each InstanceVariable was defined to hold a `ValueReference`, which also has different types. I was unsure here whether `Case` should be a subclass of a  `ValueReference` or a subclass of `InstanceVariable`. It probably doesn't make much sense to leave that comment there, though.